[그래] 로또 1단계 미션 제출합니다.

hwanghe159 opened the pull request:
피드백 부탁드립니다 감사합니다!

Commit:
docs : README 작성

Commit:
docs : README 수정

Commit:
feat : 로또 구입 금액을 입력받고 검증 후 예외처리하는 PurchasePrice 클래스 생성

Commit:
feat : 로또 구입 개수를 계산하는 메서드 추가, 출력하는 메서드 추가

Commit:
feat : 로또 한 줄을 생성하는 LottoGenerator 클래스 생성

Commit:
feat : 로또 한 줄의 정보를 담는 Lotto 클래스 생성

Commit:
feat : 로또 한 줄의 숫자를 출력하는 메서드를 뷰에 추가

Commit:
feat : 구입한 모든 로또를 관리하는 Lottos 클래스 생성

Commit:
feat : 구입한 모든 로또의 번호를 출력하는 printLottos() 생성

Commit:
docs : README 수정

당첨 번호와 보너스 번호에 대한 예외 추가 정리

Commit:
feat : 당첨 번호를 입력받아 검증하는 WinningNumbers 클래스 생성, 테스트코드 작성

Commit:
feat : 보너스 번호를 검증하는 BonusNumber 클래스 생성

Commit:
docs : README 수정

Commit:
feat : 당첨 결과를 계산하는 로직 추가

Lotto클래스에 로또 한 줄과 당첨 번호를 비교해서 비교 결과를 반환하는 메서드 추가
Lottos클래스에 로또 여러줄과 당첨 번호를 비교해서 전체 로또 결과를 반환하는 메서드 추가

Commit:
feat : 당첨 결과 계산을 위한 메서드 추가, 결과를 나타내는 MatchResult enum 생성

Commit:
test : 당첨 결과 반환 로직 테스트

Commit:
feat : 당첨 결과를 관리하고 수익률을 계산하는 LottoResult 클래스 생성

Commit:
test : 수익률에 관한 테스트 추가

Commit:
feat : 해당되는 당첨 결과가 없을 때 처리 로직 추가

Commit:
feat : 로또 당첨 결과와 수익률 출력 기능 추가

Commit:
refactor : 메서드 분리와 순서, 상수 정리

Commit:
test : 테스트 코드 상수명 변경과 필요없는 공백 삭제

Commit:
docs : README 수정

jeonghoon1107 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,37 @@
+    private static final int WINNING_NUMBERS_COUNT = 6;
+    private static final int FIVE_MATCH = 5;
+
+    private final List<Integer> numbers;
```
규칙3: 모든 원시값과 문자열을 포장한다.

hwanghe159 replied:
Ball클래스를 만들어서 포장했습니다!

jeonghoon1107 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,37 @@
+
+    public MatchResult findMatchResult(WinningNumbers winningNumbers, BonusNumber bonusNumber) {
+        int sameNumberCount = calculateSameNumberCountWith(winningNumbers);
+        if (sameNumberCount == FIVE_MATCH && bonusNumber.isIncluded(numbers)) {
```
if문에 나열되는 조건들이 많아질수록 가독성이 떨어지는데 메서드로 추출하면 명확하게 할 수 있을 것 같아요.
또한, MatchResult에서 보너스 여부를 가지고 있으면 어떨까요?

hwanghe159 replied:
보너스 여부를 추가했더니 좀 더 깔끔한 코드가 되었습니다

jeonghoon1107 commented on src/test/java/lotto/domain/LottoResultTest.java:
```diff
@@ -0,0 +1,62 @@
+
+    @DisplayName("로또 한 줄과 당첨번호를 비교했을때 올바른 당첨 결과를 반환하는지 확인")
+    @Test
+    void findMatchResultTest() {
```
parameterizedTest를 이용해도 좋을 것 같아요

hwanghe159 replied:
적용했습니다!

jeonghoon1107 reviewed (changes requested):
안녕하세요 그래!
구현 잘 하셨네요 💯 
간단한 피드백 남겼으니 확인해주세요 :)

Commit:
refactor : 로또 번호를 나타내는 Ball 클래스 생성에 따른 리팩토링

로또 번호와 관련된 모든 Integer 을 Ball로 감쌈
그에따른 재명명(BonusNumber -> BonusBall, WinningNumbers -> WinningBalls)
각 로또 번호에 대한 검증 로직을 모두 Ball 클래스 내부로 이전

Commit:
refactor : 리팩토링

1~45의 모든 Ball을 생성하는 로직을 Ball클래스로 옮겨 여러 클래스에 퍼져 있는 중복 상수(MIN_NUMBER, MAX_NUMBER 등) 통합
모든 당첨 결과를 나타내는 LottoResult를 Results로 재명명

hwanghe159 commented:
한가지 질문이 있습니다

제 코드에서는 당첨 번호를 "1,2,3,4,5,6" 이런 식으로 입력받는데
`InputView`는 말그대로 입력만 받고, 받은 `String`을 단순히 반환만 해야하는건지, 
간단한 `split`정도는 가능하고, 검증 후에 `List<Integer>`로 감싸서 반환할 수 있는건지,
아님 더 나아가서 각각의 `Integer`을 `Ball`로 감싸서 `List<Ball>`로 반환을 해도 되는건지,
아님 정해진 건 없고 각자의 재량에 맡기는 건지 궁금합니다.

View가 어느정도까지 할 수 있는건지(단순히 받은것만 반환을 해야하는지, 검증도 가능한지 등등) 에 대해서 알려주실 수 있나요?

jeonghoon1107 commented on src/main/java/lotto/LottoApplication.java:
```diff
@@ -0,0 +1,21 @@
+        Lottos lottos = new Lottos(LottoGenerator.generate(lottoCount));
+        OutputView.printLottos(lottos);
+
+        WinningBalls winningBalls = new WinningBalls(InputView.requestWinningNumbersInput());
```
하나의 WinningLotto로 보면 어떨까요?

jeonghoon1107 commented on src/main/java/lotto/domain/Results.java:
```diff
@@ -0,0 +1,55 @@
+
+    private final Map<MatchResult, Integer> lottoResults;
+
+    public Results(Lottos lottos, WinningBalls winningBalls, BonusBall bonusBall) {
```
정적팩토리 메서드만 사용해보면 어떨까요?

jeonghoon1107 commented on src/main/java/lotto/domain/WinningBalls.java:
```diff
@@ -0,0 +1,82 @@
+    private static final String WINNING_NUMBERS_DELIMITER = ",";
+    private static final int NUMBER_COUNT_PER_LOTTO = 6;
+
+    private final List<Ball> winningBalls;
```
Lotto를 대신 사용할 수 있지 않을까요?

jeonghoon1107 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.Collections;
+import java.util.List;
+
+public class Lotto {
```
로또가 가져야하는 책임 중에 duplicate check가 빠져있네요!
Lotto와 ball로 WinningBalls를 구성할 수 있을 것 같은데, 이때 함께 필요할 것 같아요!

jeonghoon1107 commented on src/main/java/lotto/domain/MatchResult.java:
```diff
@@ -0,0 +1,59 @@
+        this.prize = prize;
+    }
+
+    public static MatchResult findMatchResult(Lotto lotto, WinningBalls winningBalls, BonusBall bonusBall) {
```
Results에서 MatchResult.of를 호출해서 우승결과를 구하는 로직을 Results로 이동하면 어떨까요?

jeonghoon1107 commented on src/test/java/lotto/domain/LottoResultTest.java:
```diff
@@ -0,0 +1,85 @@
+
+    @BeforeEach
+    void setUp() {
+        List<Ball> balls1 = new ArrayList<>(Arrays.asList(
```
Arrays.asList만 사용해도 될 것 같아요

jeonghoon1107 commented on src/test/java/lotto/domain/LottoResultTest.java:
```diff
@@ -0,0 +1,85 @@
+        List<Ball> balls5 = new ArrayList<>(Arrays.asList(
+                new Ball(1), new Ball(2), new Ball(3), new Ball(43), new Ball(44), new Ball(45)));
+
+        List<Lotto> lottoBundle = new ArrayList<>();
```
Arrays.asList를 사용하면 어떨까요?

jeonghoon1107 reviewed (approved):
안녕하세요 그래!
구현 잘 하셨네요 👍 
몇가지 피드백을 남겼으니 다음단계에서 반영해주세요 :)

List<Ball>을 반환하는 것보다 split한 후에 받는 것이 좋을 것 같아요.
input, output과 도메인을 분리하는 식으로!
